plugins/outputs/websocket/websocket.go

```diff
@@ -0,0 +1,206 @@
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
+  # data_format = "influx"
+
+  ## Additional HTTP Upgrade headers
+  # [outputs.websocket.headers]
+  #   Authorization = "Bearer <TOKEN>"
+`
+
+const (
+	defaultConnectTimeout = 5 * time.Second
+	defaultWriteTimeout   = 5 * time.Second
+	defaultReadTimeout    = 30 * time.Second
+)
+
+// WebSocket can output to WebSocket endpoint.
+type WebSocket struct {
+	URL            string            `toml:"url"`
+	ConnectTimeout config.Duration   `toml:"connect_timeout"`
+	WriteTimeout   config.Duration   `toml:"write_timeout"`
+	ReadTimeout    config.Duration   `toml:"read_timeout"`
+	Headers        map[string]string `toml:"headers"`
+	UseTextFrames  bool              `toml:"use_text_frames"`
+	tls.ClientConfig
+
+	conn       *ws.Conn
+	serializer serializers.Serializer
+}
+
+// SetSerializer implements serializers.SerializerOutput.
+func (w *WebSocket) SetSerializer(serializer serializers.Serializer) {
+	w.serializer = serializer
+}
+
+// Description of plugin.
+func (w *WebSocket) Description() string {
+	return "Generic WebSocket output writer."
+}
+
+// SampleConfig returns plugin config sample.
+func (w *WebSocket) SampleConfig() string {
+	return sampleConfig
+}
+
+var errInvalidURL = errors.New("invalid websocket URL")
+
+// Connect to the output endpoint.
+func (w *WebSocket) Connect() error {
+	if parsedURL, err := url.Parse(w.URL); err != nil || (parsedURL.Scheme != "ws" && parsedURL.Scheme != "wss") {
+		return fmt.Errorf("%w: \"%s\"", errInvalidURL, w.URL)
+	}
+
+	tlsCfg, err := w.ClientConfig.TLSConfig()
+	if err != nil {
+		return fmt.Errorf("error creating TLS config: %v", err)
+	}
+
+	dialer := &ws.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: time.Duration(w.ConnectTimeout),
+		TLSClientConfig:  tlsCfg,
+	}
+
+	headers := http.Header{}
+	for k, v := range w.Headers {
+		headers.Set(k, v)
+	}
+
+	conn, resp, err := dialer.Dial(w.URL, headers)
+	if err != nil {
+		return fmt.Errorf("error dial: %v", err)
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusSwitchingProtocols {
+		return fmt.Errorf("wrong status code while connecting to server: %d", resp.StatusCode)
+	}
+
+	w.conn = conn
+	go w.read(conn)
+
+	return nil
+}
+
+func (w *WebSocket) read(conn *ws.Conn) {
+	defer func() { _ = conn.Close() }()
+	if w.ReadTimeout > 0 {
+		if err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout))); err != nil {
+			return
+		}
+		conn.SetPingHandler(func(string) error {
+			err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout)))
+			if err != nil {
+				return err
+			}
+			return conn.WriteControl(ws.PongMessage, nil, time.Now().Add(time.Duration(w.WriteTimeout)))
+		})
+	}
+	for {
+		// Need to read a connection (to properly process pings from a server).
+		_, _, err := conn.ReadMessage()
+		if err != nil {
+			return
```

Comment:
probably want to log this message. Note any returns from this function will stop the reading of the websocket and the plugin will effectively be dead. You might want to log and continue instead, and/or handle relaunching this goroutine when a reconnection happens..

Reply:
Connection is not readable after first error, it's going to error state. In the beginning of this goroutine we have defer section that closes such connection. After that connection will be tried to reestablish on next Write. Added logging for unexpected errors following example from Gorilla repo.

Reply:
Can you please add a comment to the code explaining the behavior. Now that I read your comment here it is clear to me, but before I had a hard time understanding the intended behavior.

Reply:
Done